gatewayApp/gateway/node_connection_client.py

```diff
@@ -0,0 +1,278 @@
+
+# TODO: change the DH to more secure version
+def DH_exchange(socket: socket) -> bytes:
+    """Executes Diffie-Hellman key exchange and establishes connection 
+    between Gateway and Node.
+
+    Args:
+        socket (socket): Socket for communication.
+
+    Returns:
+        bytes: AES key.
+    """
+    try:
+        socket.connect((HOST, PORT))
+        print("Connected to the server.")
+        print("Initiating DH exchange.")
+
+        # Receive the DH parameters (this is just the parameters, not a private key)
+        dh_parameters_pem: bytes = receive_data(socket)
+        print("DH parameters received")
+
+        # Load DH parameters
+        parameters = serialization.load_pem_parameters(dh_parameters_pem)
+
+        #! private key will be predetermined
+        # Generate the client's private key using the received parameters
+        client_private_key = parameters.generate_private_key()
+
+        # Generate the client's public key
+        client_public_key = client_private_key.public_key()
+        print("Key pair generated")
+
+        # Send client's public key to the server
+        client_public_key_pem: bytes = client_public_key.public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo
+        )
+        send_data(socket, client_public_key_pem)
+        print("Key pem send")
+
+        # Receive server's public key
+        server_public_key_pem: bytes = receive_data(socket)
+        server_public_key = serialization.load_pem_public_key(server_public_key_pem)
+        print("Key pem received")
+
+        # Perform Diffie-Hellman key exchange
+        shared_key: bytes = client_private_key.exchange(server_public_key)
+
+        # Derive AES key
+        aes_key: bytes = HKDF(
+            algorithm=SHA256(),
+            length=32,
+            salt=None,
+            info=b'handshake data'
+        ).derive(shared_key)
+        
+        return aes_key
+    except Exception as e:
+        print(e)
+        return None
+
+def establish_connection(node_socket: socket) -> bytes:
+    """Establishes connection with Node.
+    
+    Args:
+        node_socket (socket): Socket for communication.
+
+    Raises:
+        ConnectionError: Couldn't connect to Node.
+
+    Returns:
+        bytes: AES key.
+    """
+    for attempt in range(ATTEMPTS):
+            try:
+                aes_key = DH_exchange(node_socket)
+                if aes_key:
+                    print("Connection established.")
+                    return aes_key
+            except Exception as e:
+                print(f"Attempt {attempt + 1} failed: {e}")
+            
+            time.sleep(REFRACTORY_PERIOD)
+            
+            if attempt == ATTEMPTS - 1:
+                print("All connection attempts failed. Exiting...")
+                raise ConnectionError("Error trying to connect with Node")
+
+def process_message(message: Message) -> None:
```

Comment:
Does this function process message anyhow? Or will?
Maybe change name to print_message or sth.